src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java

```diff
@@ -119,21 +124,25 @@
     public <C extends Credentials> List<C> getCredentials(@NonNull Class<C> type, @Nullable ItemGroup itemGroup,
                                                           @Nullable Authentication authentication,
                                                           @NonNull List<DomainRequirement> domainRequirements) {
-        if (authentication == null) {
-            authentication = ACL.SYSTEM;
-        }
         List<C> result = new ArrayList<C>();
+        Set<String> ids = new HashSet<String>();
         if (ACL.SYSTEM.equals(authentication)) {
             while (itemGroup != null) {
                 if (itemGroup instanceof AbstractFolder) {
                     final AbstractFolder<?> folder = AbstractFolder.class.cast(itemGroup);
                     FolderCredentialsProperty property = folder.getProperties().get(FolderCredentialsProperty.class);
                     if (property != null) {
-                        result.addAll(DomainCredentials.getCredentials(
+                        for (C c : DomainCredentials.getCredentials(
                                 property.getDomainCredentialsMap(),
                                 type,
                                 domainRequirements,
-                                CredentialsMatchers.always()));
+                                CredentialsMatchers.always())) {
+                            if (!(c instanceof IdCredentials) || ids.add(((IdCredentials) c).getId())) {
+                                // if IdCredentials, only add if we havent added already
+                                // if not IdCredentials, always add
```

Comment:
Are there any non-`IdCredentials`, really? Should we just deprecate that possibility?